### PR TITLE
dhcp: length of each user class field must be positive

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/libsystemd-network/dhcp-option.c [hunk 1]
```diff
@@ -41,8 +41,8 @@ static int option_append(uint8_t options[], size_t size, size_t *offset,
                 STRV_FOREACH(s, (char **) optval) {
                         size_t len = strlen(*s);
 
-                        if (len > 255)
-                                return -ENAMETOOLONG;
+                        if (len > 255 || len == 0)
+                                return -EINVAL;
```
Comment (bluca) [inline src/libsystemd-network/dhcp-option.c:48]: this is an internal API only, right?

File: src/libsystemd-network/sd-dhcp-client.c [hunk 2]
```diff
@@ -576,13 +576,16 @@ int sd_dhcp_client_set_user_class(
...
+        const char * const *p;
 
-        STRV_FOREACH(p, (char **) user_class)
-                if (strlen(*p) > 255)
-                        return -ENAMETOOLONG;
+        STRV_FOREACH(p, user_class) {
+                size_t n = strlen(*p);
 
-        s = strv_copy((char **) user_class);
+                if (n > 255 || n == 0)
```
Comment (ssahani) [inline src/libsystemd-network/sd-dhcp-client.c:587]: Does strv allows 0 length strings ?

File: src/libsystemd-network/dhcp-option.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/dhcp-option.c:48]: Sorry, what do you mean?

File: src/libsystemd-network/sd-dhcp-client.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client.c:587]: Yes. So, we need to drop zero length entries explicitly.

**Comment by yuwata** [comment]
@bluca and @ssahani I added several relevant commits. PTAL.

File: src/libsystemd-network/dhcp-option.c (hunk 1 above)
Comment (ssahani) [reply src/libsystemd-network/dhcp-option.c:48]: It's a library yes

Comment (bluca) [reply src/libsystemd-network/dhcp-option.c:48]: Given the return value is changing, I was just double checking that we are not breaking some public API

File: src/libsystemd-network/sd-dhcp-client.c [hunk 3]
```diff
@@ -573,22 +573,30 @@ int sd_dhcp_client_set_mud_url(
...
 
-        STRV_FOREACH(p, (char **) user_class)
-                if (strlen(*p) > 255)
-                        return -ENAMETOOLONG;
+        assert_return(client, -EINVAL);
+
+        if (strv_isempty(user_class)) {
+                client->user_class = strv_free(client->user_class);
+                return 0;
+        }
```
Comment (ssahani) [inline src/libsystemd-network/sd-dhcp-client.c]: Which situation the user wants to drop these values. We use this for parser to reset . Reconfiguring is automatically does this by dropping the client memory.

**Review by ssahani — COMMENTED** [review]
I think mostly ok.

File: src/libsystemd-network/dhcp-option.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/dhcp-option.c:48]: sd-dhcp-client or sd-dhcp6-client are not public. So, it is safe to change that.
(NetworkManager uses these library, but I hope the change is nicely followed by the devs.)

File: src/libsystemd-network/sd-dhcp-client.c (hunk 3 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client.c]: The logic is not used by networkd. So, this change is not necessary, but does not introduce any bad behavior, I think.

Comment (ssahani) [reply src/libsystemd-network/sd-dhcp-client.c]: We should not allow zero length values . ```assert_return``` should do the job.

Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client.c]: Hmm, I thought that we may introduce some dbus or varlink method to reconfigure dhcp or dhcp6 client in the future. And when it comes, it is useful that the method accepts an empty strv and clear the user class. But ok. Such a method will not come in the near future, I guess. Let's add that clearing logic when it is actually needed.

Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client.c]: Dropped.

**Comment by ssahani** [comment]
Lgtm

**Comment by bluca** [comment]
Computer says no:

```
In file included from ../src/basic/alloc-util.h:9,
                 from ../src/libsystemd-network/sd-dhcp6-client.c:13:
../src/libsystemd-network/sd-dhcp6-client.c: In function ‘sd_dhcp6_client_set_request_vendor_class’:
../src/libsystemd-network/sd-dhcp6-client.c:494:24: error: implicit declaration of function ‘strv_isemtpy’; did you mean ‘strv_isempty’? [-Werror=implicit-function-declaration]
  494 |         assert_return(!strv_isemtpy(vendor_class), -EINVAL);
      |                        ^~~~~~~~~~~~
../src/basic/macro.h:29:42: note: in definition of macro ‘_likely_’
   29 | #define _likely_(x) (__builtin_expect(!!(x), 1))
      |                                          ^
../src/basic/macro.h:425:22: note: in expansion of macro ‘assert_log’
  425 |                 if (!assert_log(expr, #expr))                           \
      |                      ^~~~~~~~~~
../src/libsystemd-network/sd-dhcp6-client.c:494:9: note: in expansion of macro ‘assert_return’
  494 |         assert_return(!strv_isemtpy(vendor_class), -EINVAL);
      |         ^~~~~~~~~~~~~
../src/libsystemd-network/sd-dhcp6-client.c:494:24: error: nested extern declaration of ‘strv_isemtpy’ [-Werror=nested-externs]
  494 |         assert_return(!strv_isemtpy(vendor_class), -EINVAL);
      |                        ^~~~~~~~~~~~
../src/basic/macro.h:29:42: note: in definition of macro ‘_likely_’
   29 | #define _likely_(x) (__builtin_expect(!!(x), 1))
      |                                          ^
../src/basic/macro.h:425:22: note: in expansion of macro ‘assert_log’
  425 |                 if (!assert_log(expr, #expr))                           \
      |                      ^~~~~~~~~~
../src/libsystemd-network/sd-dhcp6-client.c:494:9: note: in expansion of macro ‘assert_return’
  494 |         assert_return(!strv_isemtpy(vendor_class), -EINVAL);
      |         ^~~~~~~~~~~~~
```

**Comment by yuwata** [comment]
@bluca Ouch! Fixed.

**Comment by poettering** [comment]
lgtm, too

**Pull request merged**